kafka/consumer/main.go

```diff
@@ -190,42 +212,132 @@ var datakey struct {
-	if err != nil {
-		return nil, err
-	}
-	req.Header.Set("Content-Type", "application/json")
-	resp, err := client.Do(req)
-	var bodyText []byte
-	if resp != nil && resp.Body != nil {
-		limitSize := resp.ContentLength
-		const mb134 = 1 << 27 //134MB
-		if limitSize < 1 || limitSize > mb134 {
-			limitSize = mb134
+func WithRetry(operation func() error) error {
+	backoff := initialBackoff
+	var lastErr error
+
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		err := operation()
+		if err == nil {
+			return nil // success on this attempt
 		}
-		bodyText, _ = io.ReadAll(io.LimitReader(resp.Body, int64(limitSize)))
-		resp.Body.Close()
-	}
-	if err != nil {
-		log.Printf("Error response body from SKR: %s", bodyText)
-		return nil, err
+		lastErr = err // capture for final return if out of tries
+
+		if attempt == maxRetries {
+			break // no more attempts
+		}
+
+		log.Printf("[WithRetry] attempt %d/%d failed (%v). Retrying in %s...",
+			attempt, maxRetries, err, backoff)
+
+		time.Sleep(backoff)
+		// exponential growth with clamp
+		backoff = time.Duration(math.Min(float64(backoff*2), float64(maxBackoff)))
 	}
-	if resp.StatusCode < 200 || resp.StatusCode > 207 {
-		return nil, fmt.Errorf("unable to retrieve key from SKR.  Response Code %d.  Message %s", resp.StatusCode, string(bodyText))
+
+	return fmt.Errorf("operation failed after %d retries: %w", maxRetries, lastErr)
+}
+
+func getStatus() error {
+	operation := func() error {
+		client := &http.Client{}
+		req, err := http.NewRequest("GET", "http://localhost:8080/status", nil)
+		if err != nil {
+			return fmt.Errorf("Error creating HTTP GET request: %w", err)
+		}
+		req.Header.Set("Content-Type", "application/json")
+		resp, err := client.Do(req)
+		if err != nil {
+			return fmt.Errorf("HTTP GET error from SKR: %w", err)
+		}
+
+		if resp.StatusCode < 200 || resp.StatusCode > 207 {
+			return fmt.Errorf("HTTP GET Status code not 2xx: %d.", resp.StatusCode)
+		}
+
+		var bodyText []byte
+		if resp != nil && resp.Body != nil {
+			limitSize := resp.ContentLength
+			const mb134 = 1 << 27 //134MB
+			if limitSize < 1 || limitSize > mb134 {
+				limitSize = mb134
+			}
+			bodyText, _ = io.ReadAll(io.LimitReader(resp.Body, int64(limitSize)))
+			resp.Body.Close()
+		}
+
+		log.Printf("SKR status: %s", string(bodyText))
+
+		return nil
 	}
 
-	if err := json.Unmarshal(bodyText, &datakey); err != nil {
-		log.Printf("retrieve key unmarshal error: %s", err.Error())
+	return WithRetry(operation)
+}
+
+func retrieveKey() (*rsa.PrivateKey, error) {
+	maaEndpoint := os.Getenv("SkrClientMAAEndpoint")
+	akvEndpoint := os.Getenv("SkrClientAKVEndpoint")
+	skrClientKID := os.Getenv("SkrClientKID")
+
+	var key *rsa.PrivateKey
+	log.Printf("[retrieveKey] Using environment variables:\n  SkrClientMAAEndpoint=%s\n  SkrClientAKVEndpoint=%s\n  SkrClientKID=%s", maaEndpoint, akvEndpoint, skrClientKID)
+
+	operation := func() error {
```

Comment:
should `key` be an input reference to this func? rn it's acting like a global variable